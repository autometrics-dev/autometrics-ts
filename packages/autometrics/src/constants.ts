// Spec version
export const AUTOMETRICS_VERSION = "1.0.0";

export const AUTOMETRICS_DEFAULT_SERVICE_NAME =
  "AUTOMETRICS_TYPESCRIPT_SERVICE";

// Metrics
export const COUNTER_NAME = "function.calls" as const;
export const HISTOGRAM_NAME = "function.calls.duration" as const;
export const GAUGE_NAME = "function.calls.concurrent" as const;
export const BUILD_INFO_NAME = "build_info" as const;

// Labels
export const AUTOMETRICS_VERSION_LABEL = "autometrics.version" as const;
export const BRANCH_LABEL = "branch" as const;
export const CALLER_FUNCTION_LABEL = "caller.function" as const;
export const CALLER_MODULE_LABEL = "caller.module" as const;
export const COMMIT_LABEL = "commit" as const;
export const FUNCTION_LABEL = "function" as const;
export const MODULE_LABEL = "module" as const;
export const OBJECTIVE_NAME_LABEL = "objective.name" as const;
export const OBJECTIVE_PERCENTILE_LABEL = "objective.percentile" as const;
export const OBJECTIVE_LATENCY_THRESHOLD_LABEL =
  "objective.latency_threshold" as const;
export const REPOSITORY_URL_LABEL = "repository.url" as const;
export const REPOSITORY_PROVIDER_LABEL = "repository.provider" as const;
export const RESULT_LABEL = "result" as const;
export const SERVICE_NAME_LABEL = "service.name" as const;
export const VERSION_LABEL = "version" as const;

// Descriptions
export const COUNTER_DESCRIPTION =
  "Autometrics counter for tracking function calls" as const;
export const HISTOGRAM_DESCRIPTION =
  "Autometrics histogram for tracking function call duration" as const;
export const GAUGE_DESCRIPTION =
  "Autometrics gauge for tracking concurrent function calls" as const;
export const BUILD_INFO_DESCRIPTION =
  "Autometrics info metric for tracking software version and build details" as const;

// Units
export const HISTOGRAM_UNIT = "seconds" as const;
