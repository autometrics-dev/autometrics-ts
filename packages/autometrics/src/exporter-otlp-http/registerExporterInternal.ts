import { MetricReader } from "npm:@opentelemetry/sdk-metrics@^1.17.0";

import { ExporterOptions, registerExporter } from "../../mod.ts";

export let metricReader: MetricReader | undefined;

/**
 * Makes the registered metric reader available internally for use by tests.
 *
 * This function (and the variable above) should not be exposed to consumers of
 * the library.
 *
 * @internal
 */
export function registerExporterInternal(options: ExporterOptions) {
  metricReader = options.metricReader;
  registerExporter(options);
}
