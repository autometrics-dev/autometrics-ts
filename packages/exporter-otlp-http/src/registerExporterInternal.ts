import { ExporterOptions, registerExporter } from "@autometrics/autometrics";
import { MetricReader } from "@opentelemetry/sdk-metrics";

export let metricReader: MetricReader | undefined;

/**
 * Makes the registered metric reader available internally for use by tests.
 *
 * This function (and the variable above) should not be exposed to consumers of
 * the library.
 *
 * @internal
 */
export function registerExporterInternal(options: ExporterOptions) {
  metricReader = options.metricReader;
  registerExporter(options);
}
