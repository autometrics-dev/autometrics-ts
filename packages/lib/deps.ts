export * as otelApi from "../../vendor/opentelemetry-api/mod.ts";
export * as otelCore from "../../vendor/opentelemetry-core/mod.ts";
export * as otelExporterPrometheus from "../../vendor/opentelemetry-exporter-prometheus/mod.ts";
export * as otelResources from "../../vendor/opentelemetry-resources/mod.ts";
export * as otelSdkMetrics from "../../vendor/opentelemetry-sdk-metrics/mod.ts";
export * as otelSemanticConventions from "../../vendor/opentelemetry-semantic-conventions/mod.ts";
