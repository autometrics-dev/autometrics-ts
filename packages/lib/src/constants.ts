// Metrics
export const COUNTER_NAME = "function.calls" as const;
export const HISTOGRAM_NAME = "function.calls.duration" as const;
export const GAUGE_NAME = "function.calls.concurrent" as const;
export const BUILD_INFO_NAME = "build_info" as const;

// Descriptions
export const COUNTER_DESCRIPTION =
  "Autometrics counter for tracking function calls" as const;
export const HISTOGRAM_DESCRIPTION =
  "Autometrics histogram for tracking function call duration" as const;
export const GAUGE_DESCRIPTION =
  "Autometrics gauge for tracking concurrent function calls" as const;
export const BUILD_INFO_DESCRIPTION =
  "Autometrics info metric for tracking software version and build details" as const;

// Units
export const HISTOGRAM_UNIT = "seconds" as const;
