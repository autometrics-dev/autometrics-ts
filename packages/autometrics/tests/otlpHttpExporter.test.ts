import { autometrics } from "../mod.ts";
import { COUNTER_NAME } from "../src/constants.ts";
import { init } from "../src/exporter-otlp-http/mod.ts";
import { metricReader } from "../src/exporter-otlp-http/registerExporterInternal.ts";
import { assertEquals, assertSnapshot } from "./deps.ts";

const foo = autometrics(function foo() {});

Deno.test("OTLP exporter", async (t) => {
  // make sure that metrics that are collected before `init()` is called are
  // correctly tracked.
  await t.step(
    "collects metrics recorded before init() was called",
    async () => {
      foo(); // one before

      init({ url: "http://localhost:4317", pushInterval: 5000 });

      foo(); // one after

      const collectionResult = await metricReader?.collect();

      const counterMetric =
        collectionResult?.resourceMetrics.scopeMetrics[0].metrics.find(
          (metric) => metric.descriptor.name === COUNTER_NAME,
        );

      if (!counterMetric) {
        throw new Error("Counter metric not recorded");
      }

      assertEquals(counterMetric.dataPoints[0].value, 2);

      await assertSnapshot(t, counterMetric.dataPoints[0].attributes);
    },
  );
});
