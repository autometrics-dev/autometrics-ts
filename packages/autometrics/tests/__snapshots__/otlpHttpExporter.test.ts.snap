export const snapshot = {};

snapshot[`OTLP/HTTP exporter 1`] = `
{
  "caller.function": "",
  "caller.module": "",
  "objective.name": "",
  "objective.percentile": "",
  function: "foo",
  module: "/packages/autometrics/tests/otlpHttpExporter.test.ts",
  result: "ok",
  service_name: "autometrics-monorepo",
}
`;
